src/jit/gentree.h

```diff
@@ -4037,26 +3884,198 @@ struct GenTreeIndir : public GenTreeOp
+    {
+        return (gtFlags & GTF_BLK_UNALIGNED) != 0;
+    }
+
+    // Instruction selection: during codegen time, what code sequence we will be using
+    // to encode this operation.
+    enum
+    {
+        BlkOpKindInvalid,
+        BlkOpKindHelper,
+        BlkOpKindRepInstr,
+        BlkOpKindUnroll,
+    } gtBlkOpKind;
+
+    bool gtBlkOpGcUnsafe;
+
+    GenTreeBlk(genTreeOps oper, var_types type, GenTreePtr addr, unsigned size)
+        : GenTreeIndir(oper, type, addr, nullptr)
+        , gtBlkSize(size)
+        , gtBlkOpKind(BlkOpKindInvalid)
+        , gtBlkOpGcUnsafe(false)
+    {
+        assert(OperIsBlk(oper));
+        gtFlags |= (addr->gtFlags & GTF_ALL_EFFECT);
+    }
+
+    GenTreeBlk(genTreeOps oper, var_types type, GenTreePtr addr, GenTreePtr data, unsigned size)
+        : GenTreeIndir(oper, type, addr, data), gtBlkSize(size), gtBlkOpKind(BlkOpKindInvalid), gtBlkOpGcUnsafe(false)
+    {
+        assert(OperIsBlk(oper));
+        gtFlags |= (addr->gtFlags & GTF_ALL_EFFECT);
+        gtFlags |= (data->gtFlags & GTF_ALL_EFFECT);
+    }
+
+#if DEBUGGABLE_GENTREE
+protected:
+    friend GenTree;
+    GenTreeBlk() : GenTreeIndir()
+    {
+    }
+#endif // DEBUGGABLE_GENTREE
+};
+
+// gtObj  -- 'object' (GT_OBJ).
+//
+// This node is used for block values that may have GC pointers.
 
-struct GenTreeObj : public GenTreeIndir
+struct GenTreeObj : public GenTreeBlk
 {
     CORINFO_CLASS_HANDLE gtClass; // the class of the object
 
-    GenTreeObj(var_types type, GenTreePtr addr, CORINFO_CLASS_HANDLE cls)
-        : GenTreeIndir(GT_OBJ, type, addr, nullptr), gtClass(cls)
+    // If non-null, this array represents the gc-layout of the class.
+    // This may be simply copied when cloning this node, because it is not changed once computed.
+    BYTE* gtGcPtrs;
+
+    // If non-zero, this is the number of slots in the class layout that
+    // contain gc-pointers.
+    __declspec(property(get = GetGcPtrCount)) unsigned gtGcPtrCount;
+    unsigned GetGcPtrCount() const
+    {
+        assert(_gtGcPtrCount != UINT32_MAX);
+        return _gtGcPtrCount;
+    }
+    unsigned _gtGcPtrCount;
+
+    // If non-zero, the number of pointer-sized slots that constitutes the class token.
+    unsigned gtSlots;
+
+    bool IsGCInfoInitialized()
+    {
+        return (_gtGcPtrCount != UINT32_MAX);
+    }
+
+    void SetGCInfo(BYTE* gcPtrs, unsigned gcPtrCount, unsigned slots)
+    {
+        gtGcPtrs      = gcPtrs;
+        _gtGcPtrCount = gcPtrCount;
+        gtSlots       = slots;
+        if (gtGcPtrCount != 0)
+        {
+            // We assume that we cannot have a struct with GC pointers that is not a multiple
+            // of the register size.
+            // The EE currently does not allow this, but it could change.
+            // Let's assert it just to be safe.
+            noway_assert(roundUp(gtBlkSize, REGSIZE_BYTES) == gtBlkSize);
+        }
+        else
+        {
+            // ChangeOper(GT_BLK);
+        }
```

Comment:
Ah, yes - I decided to eliminate that, since it's not really necessary. I will remove it.

---
In reply to: [77271784](https://github.com/dotnet/coreclr/pull/7027#discussion_r77271784) [](ancestors = 77271784)